### ISSUE TITLE
Cannot read property 'getBoundingClientRect' of null

### ISSUE THREAD

**Issue by falconmick** [dazld/react-on-visible#11]
It looks like the code that renderers the element that wraps the children passed to OnVisisble is null for a split second when I navigate pages. Don't know what else to tell you other than I have wrapped OnVisible with another pure function component that adds default properties. Would love your input

Code that sets holder to null:

    key: 'render',
    value: function render() {
    var _this3 = this;

    var visible = this.state.visible;

    var classes = (0, _classnames2.default)(this.props.className, _defineProperty({}, this.props.visibleClassName || 'visible', visible));

    return _react2.default.createElement(
        'div',
        {
            style: this.props.style,
            className: classes,
            ref: function ref(el) {
                _this3.holder = el;
            }
        },
        this.props.children
      );
    }

My Wrapper:

```
export const OnVisibleFadeDrop = props => {
  const { children, className } = props;
  return (
    <OnVisible className={classNames('on-vis-fade-drop', className, onVisibleClass)} percent={10}>{children}</OnVisible>
  );
};

```

Cheers

**Comment by dazld** [comment]
landed and published as `1.2.0`. thanks again!

**Comment by dazld** [comment]
and next - the `ref` function can be called with null, which is when the element is unmounted, letting you clean up event handlers. this should be the correct fix.

> React will call the ref callback with the DOM element when the component mounts, and call it with null when it unmounts.

https://reactjs.org/docs/refs-and-the-dom

**Comment by dazld** [comment]
ie, this is most likely buggy code in r-o-v and should be better. reopening to fix this correctly soon™

**Comment by falconmick** [comment]
Yeah I remember reading something about how they recommend you do ref changing and this was he old way or somthing

**Comment by falconmick** [comment]
![image](https://user-images.githubusercontent.com/713449/31476103-50f92e62-af37-11e7-8bff-aac1a5ae1c3e.png)
https://reactjs.org/docs/refs-and-the-dom.html#adding-a-ref-to-a-dom-element
Basically what is happening:

1. Component mounts and sets the correct ref, everybody is happy!
2. Stuff
4. The timeout schedules call for in 10ms or whenever chrome decides to actually do it
3. Component is being demounted, as par the doco shown above it is called again and null is provided.
5. even though the component is unmounted, we only deregistered the next call to setup a callback, therefore the callback is fired.

What we need to do is add a cancelation token to our callback so that on demounting we can cancel the next callback.

I have an open PR at the moment but I won't be adding this as I want to keep it pure.

**Comment by falconmick** [comment]
this could be a solution, I will try it upon completion of my other PR I will try protomixing a cancellation function: https://stackoverflow.com/questions/23020290/javascript-add-methods-to-function-prototype

**Comment by falconmick** [comment]
Upon completion of other PR, I have a fix: https://github.com/falconmick/react-on-visible/pull/2
